packages/astro/src/cli/index.ts

```diff
@@ -147,29 +149,43 @@ async function runCommand(cmd: string, flags: yargs.Arguments) {
 		case 'dev': {
 			async function startDevServer() {
 				const { watcher, stop } = await devServer(astroConfig, { logging, telemetry });
-
-				watcher.on('change', logRestartServerOnConfigChange);
-				watcher.on('unlink', logRestartServerOnConfigChange);
-				function logRestartServerOnConfigChange(changedFile: string) {
-					if (userConfigPath === changedFile) {
-						warn(logging, 'astro', 'Astro config updated. Restart server to see changes!');
-					}
-				}
-
-				watcher.on('add', async function restartServerOnNewConfigFile(addedFile: string) {
-					// if there was not a config before, attempt to resolve
-					if (!userConfigPath && addedFile.includes('astro.config')) {
-						const addedConfig = await openConfig({ cwd: root, flags, cmd, logging });
-						if (addedConfig.userConfigPath) {
-							info(logging, 'astro', 'Astro config detected. Restarting server...');
-							astroConfig = addedConfig.astroConfig;
-							userConfig = addedConfig.userConfig;
-							userConfigPath = addedConfig.userConfigPath;
+				let restartInFlight = false;
+				const configFlag = resolveFlags(flags).config;
+				const configFlagPath = configFlag
+					? (await resolveConfigURL({ cwd: root, flags }))?.pathname
+					: undefined;
+				const resolvedRoot = appendForwardSlash(resolveRoot(root));
+
+				const handleServerRestart = (logMsg: string) =>
+					async function (changedFile: string) {
+						if (
+							!restartInFlight &&
+							(configFlag
+								? // If --config is specified, only watch changes for this file
+								  configFlagPath && normalizePath(configFlagPath) === normalizePath(changedFile)
+								: // Otherwise, watch for any astro.config.* file changes in project root
+								  new RegExp(
+										`${normalizePath(resolvedRoot)}.*astro\.config\.((mjs)|(cjs)|(js)|(ts))$`
+								  ).test(normalizePath(changedFile)))
```

Comment:
(aggressive `normalizePath` usage so we're always comparing apples-to-apples file paths)